BUG: Adding extra text in html of summary2.Summary #8663

lanzariel opened the pull request:
- [ ] closes #8663
- [ ] code/documentation is well formatted.  
- [ ] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message).

josef-pkt commented:
style check fails with `statsmodels/iolib/summary2.py:204:80: E501 line too long (84 > 79 characters)`
test failures are unrelated to this PR

josef-pkt commented:
looks fine in an example.
This does not close #8663 because there is the additional problem missing extra text in model summary2

josef-pkt commented:
as explanation:
`summary2` keeps extra_text as a list of strings
`summary` uses only one string, added text is concatenated with newline `"\n"`

lanzariel commented:
You are right. I have fixed `summary2`. Now I see the difference between parsing lists and strings.

I have noticed the file `lint.sh` that uses the appropriate parameters for  `flake8 ` sorry for not noticing it before.

I have seen that other lines in `regression.linear_model` had formatting problems and I fixed them in this commit. I know it should be avoided to have many fixes in one commit, but they are minor issues I hope.

josef-pkt commented on statsmodels/regression/linear_model.py:
```diff
@@ -2637,9 +2644,10 @@ def get_robustcov_results(self, cov_type='HC1', use_t=None, **kwargs):
+            res.cov_params_default = \
+                sw.cov_nw_panel(self, maxlags, groupidx,
+                                weights_func=weights_func,
+                                use_correction=use_correction)
```
that's the only formatting change that is annoying, especially because it's not any standard formatting either.

josef-pkt commented on statsmodels/regression/linear_model.py:
```diff
@@ -2912,16 +2921,38 @@ def summary2(
+            warn = "* The condition number is large (%.g). This might \
+                indicate strong multicollinearity or other numerical \
+                problems." % condno
+            etext.append(warn)
```
you could add an `add_text` method to Summary 2 class, so we don't have to know here whether it's string or list concatenation

josef-pkt replied:
correction: there is already an add_text method

lanzariel replied:
The issue here is whether we are sure that every annotation will always be in the form
```
Notes:
[1] Note1
[2] Note2
...
```

josef-pkt commented:
I will look at it tomorrow.
If I make quick comments now, then they may not be correct because I'm not familiar enough with the code and don't have time to fully read right now.

josef-pkt commented:
The unit test TestDynamicFactor_ar2_errors is fragile and fails quite often on one of the test environments independently of PR content.

josef-pkt commented on statsmodels/regression/linear_model.py:
```diff
@@ -2912,16 +2924,38 @@ def summary2(
-            warn = "* The condition number is large (%.g). This might indicate \
-            strong multicollinearity or other numerical problems." % condno
-            smry.add_text(warn)
+            warn = "* The condition number is large (%.g). This might \
```
There is an extra star "*" in front that is not in the `summary` version

lanzariel replied:
I wonder where I got the star and the `(%.g)` part from. It should have been `%6.3g`. Sorry for the confusion.

josef-pkt commented:
There's a bug on line 2908, eigvals should not be resorted. The warning message assumes that the last entry is the smallest eigenvalue. This resorting needs to be removed (to follow the same pattern as `summary`
`        # eigvals = np.sort(eigvals)  # in increasing order`

josef-pkt commented on statsmodels/regression/linear_model.py:
```diff
@@ -2912,16 +2924,38 @@ def summary2(
             matrix is singular." % eigvals[-1]
-            smry.add_text(warn)
+            etext.append(warn)
         if condno > 1000:
```
In `summary` I put this in and `elif`. If we already have a very small eigenvalue, then we don't need to warn additionally about the condition number.

josef-pkt commented:
BTW: I'm working with an OLS example that has one of the x columns twice, so the result is singular and we should get warnings in extra_text.

josef-pkt commented:
AFAICS you still need to remove the resorting of eigvals, i.e. delete line
`        eigvals = np.sort(eigvals)  # in increasing order`

lanzariel commented:
You are right.
Also, I considered adding a notebook, which would probably be more effective than specific tests to have a quick view of what is going on, but I have some problems displaying the documentation locally. I'll do it in the next few days if I have more time.

josef-pkt commented:
I'm using a notebook to check.
unit test would be good just to check that some words are in the summaries, but it needs a singular case, which I guess we don't have in the unit tests.

When everything works for the singular case, then I will merge and it will make it into 0.14

josef-pkt commented:
I tried the latest version of the code.
Looks good in my notebook. merging

Thanks for the pull request
